libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridFromSplines.cpp

```diff
@@ -592,11 +643,335 @@ namespace meshkernel
+            {
+                continue;
+            }
+
+            Point x1 = activeLayerPoints[i];
+            Point x2 = activeLayerPoints[i + 1];
+
+            Point v1 = velocityVectorAtGridPoints[i];
+            Point v2 = velocityVectorAtGridPoints[i + 1];
+
+            const double dL1 = ComputeDistance(x1, activeLayerPoints[i + 1], m_splines->m_projection);
+            UInt indexLeft;
+            UInt indexLeftOfLeft;
+            UInt indexRight;
+            UInt indexRightOfRight;
+            UInt dummy;
+
+            std::tie(indexLeft, dummy) = GetNeighbours(activeLayerPoints, i);
+            std::tie(dummy, indexRight) = GetNeighbours(activeLayerPoints, i + 1);
+
+            std::tie(indexLeftOfLeft, dummy) = GetNeighbours(activeLayerPoints, indexLeft);
+            std::tie(dummy, indexRightOfRight) = GetNeighbours(activeLayerPoints, indexRight);
+
+            for (UInt j = 0; j < frontGridPoints.size() - 1; ++j)
+            {
+                if (!frontGridPoints[j].IsValid() || !frontGridPoints[j + 1].IsValid())
+                {
+                    continue;
+                }
+
+                Point x3 = frontGridPoints[j];
+                Point x4 = frontGridPoints[j + 1];
+
+                Point v3 = frontVelocities[j];
+                Point v4 = frontVelocities[j + 1];
+
+                const double dL2 = ComputeDistance(x3, x4, m_splines->m_projection);
+
+                double d1 = ComputeDistance(x1, x3, m_splines->m_projection);
+                double d2 = ComputeDistance(x2, x3, m_splines->m_projection);
+                double d3 = ComputeDistance(x1, x4, m_splines->m_projection);
+                double d4 = ComputeDistance(x2, x4, m_splines->m_projection);
+
+                if (d1 < tolerance || d2 < tolerance || d3 < tolerance || d4 < tolerance)
+                {
+                    continue;
+                }
+
+                if (!IncludeDirectNeighbours(layerIndex, j, gridPointsIndices, indexLeftOfLeft, indexRightOfRight))
+                {
+                    continue;
+                }
+
+                double dmin = std::min({d1, d2, d3, d4});
+
+                // get a lower bound for the cross time
+                double hlow2 = 0.25 * std::max(dmin * dmin - std::pow(0.5 * std::max(dL1, dL2), 2), 0.0);
+
+                // check if the lower bounds is larger than the minimum found so far
+                double vv1 = std::sqrt(length(v3 - v1));
+                double vv2 = std::sqrt(length(v3 - v2));
+                double vv3 = std::sqrt(length(v4 - v1));
+                double vv4 = std::sqrt(length(v4 - v2));
+
+                const double maxvv = std::max(std::max(vv1, vv2), std::max(vv3, vv4));
+
+                if (std::sqrt(hlow2) > maxvv * std::min(tmax[i], tmax[i + 1]))
+                {
+                    continue;
+                }
+
+                double t1 = ComputeNodeSegmentCrossingTime(x1, x3, x4, v1, v3, v4);
```

Comment:
// Compute crossing times
            const std::array<double, 4> crossingTimes = {
                ComputeNodeSegmentCrossingTime(x1, x3, x4, v1, v3, v4),
                ComputeNodeSegmentCrossingTime(x2, x3, x4, v2, v3, v4),
                ComputeNodeSegmentCrossingTime(x3, x1, x2, v3, v1, v2),
                ComputeNodeSegmentCrossingTime(x4, x1, x2, v4, v1, v2)
            };

            const double tmin = *std::min_element(crossingTimes.begin(), crossingTimes.end());

Reply:
I would guess that this version will be slower than the current code